app/models/work.rb

```diff
@@ -292,13 +292,19 @@ def new_notification_count_for_user(user_id)
   # In practice, the user_id is the id of the current user and therefore this method marks the current's user
   # notifications as read.
   def mark_new_notifications_as_read(user_id)
-    activities.each do |activity|
-      unread_notifications = WorkActivityNotification.where(user_id:, work_activity_id: activity.id, read_at: nil)
-      unread_notifications.each do |notification|
-        notification.read_at = Time.now.utc
-        notification.save
-      end
-    end
```

Comment:
This was culprit number 1. We iterated over all activities for the work and then we issued another SQL SELECT to fetch unread notifications, so even when there were no unread notifications we were iterating individually. The new code does both at once (SQL SELECT and UPDATE) in batches of 1000 records.